go/types/graph_builder.go

```diff
@@ -0,0 +1,301 @@
+	b := &GraphBuilder{oc: opc, vrw: vrw, verbose: verbose}
+	b.pushNewKeyOnStack(String("ROOT"), rootKind)
+	return b
+}
+
+// MapSet() will add the key/value pair |k, v| to the map found by traversing
+// the graph using the |keys| slice. Intermediate maps referenced by |keys| are
+// created as necessary. This is threadsafe, may call from multiple go routines.
+func (b *GraphBuilder) MapSet(keys []Value, k Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call MapSet() again after Build()")
+	b.oc.GraphMapSet(keys, k, v)
+}
+
+// SetInsert() will insert the value |v| into the set at path |keys|. Intermediate
+// maps referenced by |keys| are created as necessary. This is threadsafe, may
+// call from multiple go routines.
+func (b *GraphBuilder) SetInsert(keys []Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call SetInsert() again after Build()")
+	b.oc.GraphSetInsert(keys, v)
+}
+
+// ListAppends() will append |v| to the list at path |p|. Intermediate
+// maps referenced by |keys| are created as necessary. This is threadsafe, may
+// call from multiple go routines, however append semantics are such that the
+// elements will be appended in order that functions are called, so order has
+// to be managed by caller.
+func (b *GraphBuilder) ListAppend(keys []Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call ListAppend() again after Build()")
+	b.oc.GraphListAppend(keys, v)
+}
+
+type GraphOpContainer struct {
+	keys []Value
+	kind NomsKind
+	item sequenceItem
+}
+
+// Builds and returns the graph.
+func (b *GraphBuilder) Build() Value {
+	var opc opCache
+
+	checkFirstCall := func() {
+		b.mutex.Lock()
+		defer b.mutex.Unlock()
+
+		d.PanicIfTrue(b.oc == nil, "Can only call Build() once")
+		opc = b.oc
+		b.oc = nil
+	}
+
+	checkFirstCall()
+
+	iter := opc.NewIterator()
+	defer iter.Release()
+	keyCnt := int64(0)
+
+	// start up a go routine that will do the reading from graphBuilder's private
+	// ldb opCache.
+	graphOpChan := make(chan GraphOpContainer, 512)
+	go func() {
+		for iter.Next() {
+			keys, kind, item := iter.GraphOp()
+			container := GraphOpContainer{keys: keys, kind: kind, item: item}
+			graphOpChan <- container
+		}
+		close(graphOpChan)
+	}()
+
+	// iterator returns keys, in sort order by array
+	for goc := range graphOpChan {
+		keys, kind, item := goc.keys, goc.kind, goc.item
+
+		// Get index of first key that is different than what is on the stack
+		idx := commonPrefixCount(b.stack, keys)
+
+		if idx == -1 {
+			// no keys have changed we're working on same coll as previous
+			// iteration, just append to sequenceChunker at top of stack
+			b.appendItemToCurrentTopOfStack(kind, item)
+		} else {
+			// Some keys that were in the last graphOp are no longer present
+			// which indicates that we are finished adding to those cols. Pop
+			// those keys from the stack. This will cause any popped cols to be
+			// closed and added to their parents.
+			for idx < b.stack.len()-1 {
```

Comment:
you do `b.stack.len()-1` so much that I wonder if you should replace graphStack.len() with some other method that returns that value. `lastIdx()` or something?